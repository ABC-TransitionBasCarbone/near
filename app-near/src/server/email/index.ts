import { env } from "~/env";
import { type BrevoBody } from "~/types/Brevo";

const sendEmail = async (body: BrevoBody) => {
  if (env.NODE_ENV === "production") {
    return fetch("https://api.brevo.com/v3/smtp/email", {
      method: "POST",
      headers: {
        Accept: "application/json",
        "api-key": env.BREVO_API_KEY,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
  }

  console.log(`
Email send with command:
curl --request POST \
     --url https://api.brevo.com/v3/smtp/email \
     --header 'accept: application/json' \
     --header 'api-key: <fill-with-correct-api-key>' \
     --header 'content-type: application/json' \
     --data '${JSON.stringify(body).replace(/'/g, "&apos;")}'
    `);
  return `Email template ${body.templateId} send`;
};

const EmailService = { sendEmail };

export default EmailService;
