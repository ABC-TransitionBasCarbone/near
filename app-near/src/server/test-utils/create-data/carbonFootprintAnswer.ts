import { BroadcastChannel, type CarbonFootprintAnswer } from "@prisma/client";
import { faker } from "@faker-js/faker";
import { type BuilderCarbonFootprintAnswer } from "~/types/CarbonFootprint";

export const buildCarbonFootprintAnswer = (
  surveyId: number,
  suId: number,
  data?: Partial<CarbonFootprintAnswer>,
): BuilderCarbonFootprintAnswer => ({
  broadcastChannel: faker.helpers.arrayElement(Object.values(BroadcastChannel)),
  email: Date.now() + faker.internet.email(),
  globalNote: faker.number.float(),
  transportation: faker.number.float(),
  transportationCar: faker.number.float(),
  transportationCarKm2: faker.number.float(),
  transportationCarOilConsumption100: faker.number.float(),
  transportationCarOilType: faker.number.float(),
  transportationCarElectricConsumption100: faker.number.float(),
  transportationPlane: faker.number.float(),
  transportationBicycle: faker.number.float(),
  transportationSoftMobility: faker.number.float(),
  transportationTrain: faker.number.float(),
  transportationPublicTransport: faker.number.float(),
  transportationCarLongDistance: faker.number.float(),
  transportationHollidays: faker.number.float(),
  transportationHollidaysCaravan: faker.number.float(),
  transportationHollidaysCampingCar: faker.number.float(),
  transportationHollidaysVan: faker.number.float(),
  transportationHollidaysRentalVehicle: faker.number.float(),
  transportationHollidaysCruise: faker.number.float(),
  transportationFerry: faker.number.float(),
  alimentation: faker.number.float(),
  alimentationMeals: faker.number.float(),
  alimentationLunchDinner: faker.number.float(),
  alimentationBonus: faker.number.float(),
  alimentationLocalImpact: faker.number.float(),
  alimentationLocalConsumption: faker.number.float(),
  alimentationSeasonalImpact: faker.number.float(),
  alimentationAnnualBreakfast: faker.number.float(),
  alimentationDeforestation: faker.number.float(),
  alimentationDrinks: faker.number.float(),
  alimentationHotDrinks: faker.number.float(),
  alimentationColdDrinks: faker.number.float(),
  alimentationWaste: faker.number.float(),
  alimentationWasteHabits: faker.number.float(),
  logement: faker.number.float(),
  logementConstruction: faker.number.float(),
  logementConstructionBase: faker.number.float(),
  logementConstructionCooling: faker.number.float(),
  logementConstructionRenovation: faker.number.float(),
  logementConstructionDeforestation: faker.number.float(),
  logementConstructionSoilArtificialization: faker.number.float(),
  logementElectricity: faker.number.float(),
  logementElectricityTotalConsumption: faker.number.float(),
  logementElectricitySolarAutoconsumption: faker.number.float(),
  logementElectricityCarbonIntensity: faker.number.float(),
  logementHeating: faker.number.float(),
  logementAirConditioning: faker.number.float(),
  logementSwimmingPool: faker.number.float(),
  logementOutdor: faker.number.float(),
  logementHollidays: faker.number.float(),
  logementHollidaysHotelImpact: faker.number.float(),
  logementHollidaysCampingImpact: faker.number.float(),
  logementHollidaysYouthHostelImpact: faker.number.float(),
  logementHollidaysRentalImpact: faker.number.float(),
  logementHollidaysHomeExchangeImpact: faker.number.float(),
  logementHollidaysSecondaryHomeImpact: faker.number.float(),
  divers: faker.number.float(),
  diversHouseholdAppliances: faker.number.float(),
  diversHouseholdAppliancesDevices: faker.number.float(),
  diversFurniture: faker.number.float(),
  diversFurnitureItems: faker.number.float(),
  diversFurnitureDeforestation: faker.number.float(),
  diversFurniturePreservation: faker.number.float(),
  diversDigital: faker.number.float(),
  diversDigitalInternet: faker.number.float(),
  diversDigitalDevices: faker.number.float(),
  diversTextile: faker.number.float(),
  servicesSocietal: faker.number.float(),
  servicesPublics: faker.number.float(),
  servicesMarket: faker.number.float(),
  answers: { anything: "ok" },

  airTravelFrequency: null,
  digitalIntensity: null,
  heatSource: null,
  meatFrequency: null,
  purchasingStrategy: null,
  transportationMode: null,

  distanceToBarycenter: null,
  knowSu: true,

  ...data,
  surveyId,
  suId,
});
