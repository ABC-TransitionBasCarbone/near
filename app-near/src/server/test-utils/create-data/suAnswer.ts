import {
  AgeCategory,
  AirTravelFrequency,
  BroadcastChannel,
  DigitalIntensity,
  EasyHealthAccess,
  Gender,
  HeatSource,
  MeatFrequency,
  ProfessionalCategory,
  PurchasingStrategy,
  type SuAnswer,
  TransportationMode,
} from "@prisma/client";
import { faker } from "@faker-js/faker";
import { type BuilderSuAnswer } from "~/types/SuAnswer";

export const buildSuAnswer = (
  surveyId: number,
  data: Partial<SuAnswer>,
): BuilderSuAnswer => ({
  isNeighborhoodResident: faker.datatype.boolean(),
  ageCategory: faker.helpers.arrayElement(Object.values(AgeCategory)),
  airTravelFrequency: faker.helpers.arrayElement(
    Object.values(AirTravelFrequency),
  ),
  broadcastChannel: faker.helpers.arrayElement(Object.values(BroadcastChannel)),
  digitalIntensity: faker.helpers.arrayElement(Object.values(DigitalIntensity)),
  easyHealthAccess: faker.helpers.arrayElement(Object.values(EasyHealthAccess)),
  email: faker.internet.email(),
  gender: faker.helpers.arrayElement(Object.values(Gender)),
  heatSource: faker.helpers.arrayElement(Object.values(HeatSource)),
  meatFrequency: faker.helpers.arrayElement(Object.values(MeatFrequency)),
  professionalCategory: faker.helpers.arrayElement(
    Object.values(ProfessionalCategory),
  ),
  purchasingStrategy: faker.helpers.arrayElement(
    Object.values(PurchasingStrategy),
  ),
  transportationMode: faker.helpers.arrayElement(
    Object.values(TransportationMode),
  ),
  ...data,
  surveyId,
});
