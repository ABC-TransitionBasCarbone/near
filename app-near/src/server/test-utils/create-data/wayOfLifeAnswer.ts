import {
  ActionWhenTooCold,
  ActionWhenTooHot,
  AgeCategory,
  AirTravelFrequency,
  BroadcastChannel,
  DigitalIntensity,
  EngagmentDomains,
  FoodFrequency,
  Gender,
  GiveFreeTimeToHelp,
  HeatSource,
  HomeOccupationType,
  HowManyPeopleCanIHelp,
  MeatFrequency,
  ParksUsageFrequency,
  ProfessionalCategory,
  PurchasingStrategy,
  ReasonsToEatMeat,
  ReasonsToNotBuyFrenchSeasonFood,
  ReasonsToNotChoseSecondHand,
  ReasonsUsingCar,
  TransportationMode,
  TransportMode,
  TransportModeToStation,
  TransportModeToTravel,
  TransportTime,
  TypeOfFood,
  WishesChoices,
  YesNo,
  type WayOfLifeAnswer,
} from "@prisma/client";
import { faker } from "@faker-js/faker";
import { type BuilderWayOfLifeAnswer } from "~/types/WayOfLifeAnswer";

export const buildWayOfLifeAnswer = (
  surveyId: number,
  data?: Partial<WayOfLifeAnswer>,
): BuilderWayOfLifeAnswer => ({
  accessToFoodServiceSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  accessToShortFoodCircuitSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  actionWhenTooCold: faker.helpers.arrayElements(
    Object.values(ActionWhenTooCold),
  ),
  actionWhenTooHot: faker.helpers.arrayElements(
    Object.values(ActionWhenTooHot),
  ),
  ageCategory: faker.helpers.arrayElement(Object.values(AgeCategory)),
  airTravelFrequency: faker.helpers.arrayElement(
    Object.values(AirTravelFrequency),
  ),
  associativeActivity: faker.helpers.arrayElement(Object.values(YesNo)),
  bicycleRepairShopSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  capacityToShareIdeaToTownHall: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  carAnPedestriansRespect: faker.helpers.arrayElement(Object.values(YesNo)),
  classicFoodMarketFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  clothesRepairShopSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  comment: faker.lorem.sentence(),
  confortHouseWhenHot: faker.helpers.arrayElement(Object.values(YesNo)),
  culturalActivity: faker.helpers.arrayElement(Object.values(YesNo)),
  digitalIntensity: faker.helpers.arrayElement(Object.values(DigitalIntensity)),
  easyBicycle: faker.helpers.arrayElement(Object.values(YesNo)),
  easyPublicTransports: faker.helpers.arrayElement(Object.values(YesNo)),
  easyToLeaveCityWithTranports: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  easyWalking: faker.helpers.arrayElement(Object.values(YesNo)),
  electronicRepairShopSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  foodMarketZone: faker.lorem.word(),
  gender: faker.helpers.arrayElement(Object.values(Gender)),
  giveFreeTimeToHelp: faker.helpers.arrayElement(
    Object.values(GiveFreeTimeToHelp),
  ),
  groceryStoreFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  hardDiscountFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  heatSource: faker.helpers.arrayElement(Object.values(HeatSource)),
  hobbiesSpaces: faker.helpers.arrayElement(Object.values(YesNo)),
  hobbyZone: faker.lorem.word(),
  homeDeliveryFoodStoreFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  homeOccupationType: faker.helpers.arrayElement(
    Object.values(HomeOccupationType),
  ),
  howManyPeopleCanIHelp: faker.helpers.arrayElement(
    Object.values(HowManyPeopleCanIHelp),
  ),
  hypermarketFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  ideaEasyTalk: faker.helpers.arrayElement(Object.values(YesNo)),
  knowSu: faker.helpers.arrayElement([true, false]),
  localShopsToMeetYourNeeds: faker.helpers.arrayElement(Object.values(YesNo)),
  meatFrequency: faker.helpers.arrayElement(Object.values(MeatFrequency)),
  neighborhoodLife: faker.helpers.arrayElement(Object.values(YesNo)),
  neighborhoodOrganicMarketSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  neighborhoodOrganicProductsSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  neighborhoodSeasonFruitAndVegetablesSatisfaction: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  noInformationOnCitizenParticipation: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  notColdHouse: faker.helpers.arrayElement(Object.values(YesNo)),
  notTooMuchTraffic: faker.helpers.arrayElement(Object.values(YesNo)),
  organicOrSolidarityFoodStoreFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  otherFoodFrequencyInformation: faker.lorem.sentence(),
  otherFoodSatisfactionInformation: faker.lorem.sentence(),
  otherHousingInformation: faker.lorem.sentence(),
  otherLocalPoliticInformation: faker.lorem.sentence(),
  otherMutualAidInformation: faker.lorem.sentence(),
  otherNeighborhoodLifeInformation: faker.lorem.sentence(),
  otherParksInformation: faker.lorem.sentence(),
  otherRepairShopSatisfactionInformation: faker.lorem.sentence(),
  otherServicesInformation: faker.lorem.sentence(),
  otherTransportationInformation: faker.lorem.sentence(),
  parksUsageFrequency: faker.helpers.arrayElement(
    Object.values(ParksUsageFrequency),
  ),
  preferBuyFrenchAndSeasonFood: faker.helpers.arrayElement(
    Object.values(WishesChoices),
  ),
  preferSecondHand: faker.helpers.arrayElement(Object.values(WishesChoices)),
  privateOrShareFieldToFarm: faker.helpers.arrayElement(Object.values(YesNo)),
  professionalCategory: faker.helpers.arrayElement(
    Object.values(ProfessionalCategory),
  ),
  publicServicesPresence: faker.helpers.arrayElement(Object.values(YesNo)),
  purchasingStrategy: faker.helpers.arrayElement(
    Object.values(PurchasingStrategy),
  ),
  reasonsToContinueUsingCar: faker.helpers.arrayElements(
    Object.values(ReasonsUsingCar),
  ),
  reasonsToEatMeat: faker.helpers.arrayElements(
    Object.values(ReasonsToEatMeat),
  ),
  reasonsToNotBuyFrenchAndSeasonFood: faker.helpers.arrayElements(
    Object.values(ReasonsToNotBuyFrenchSeasonFood),
  ),
  reasonsToNotChoseSecondHand: faker.helpers.arrayElements(
    Object.values(ReasonsToNotChoseSecondHand),
  ),
  remoteWorkingWeeklyFrequency: faker.helpers.arrayElement([
    "1",
    "2",
    "3",
    "4",
    "5",
  ]),
  secondHandShopSatisfaction: faker.helpers.arrayElement(Object.values(YesNo)),
  servicesToShareOrRentObjects: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  specializedFoodStoreFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  supermarketFrequency: faker.helpers.arrayElement(
    Object.values(FoodFrequency),
  ),
  transportationMode: faker.helpers.arrayElement(
    Object.values(TransportationMode),
  ),
  transportModeToBuyFood: faker.helpers.arrayElement(
    Object.values(TransportMode),
  ),
  transportModeToGoToStation: faker.helpers.arrayElement(
    Object.values(TransportModeToStation),
  ),
  transportModeToHobby: faker.helpers.arrayElement(
    Object.values(TransportMode),
  ),
  transportModeToTravel: faker.helpers.arrayElement(
    Object.values(TransportModeToTravel),
  ),
  transportModeToWork: faker.helpers.arrayElement(Object.values(TransportMode)),
  transportTimeToBuyFood: faker.helpers.arrayElement(
    Object.values(TransportTime),
  ),
  transportTimeToHobby: faker.helpers.arrayElement(
    Object.values(TransportTime),
  ),
  transportTimeToWork: faker.helpers.arrayElement(Object.values(TransportTime)),
  typeOfFood: faker.helpers.arrayElement(Object.values(TypeOfFood)),
  vegetalParksSatisfaction: faker.helpers.arrayElement(Object.values(YesNo)),
  wantToParticipateToCivicInitiatives: faker.helpers.arrayElement(
    Object.values(YesNo),
  ),
  wantToReduceCarUsage: faker.helpers.arrayElement(
    Object.values(WishesChoices),
  ),
  wantToReduceMeatAndFish: faker.helpers.arrayElement(
    Object.values(WishesChoices),
  ),
  workZone: faker.lorem.word(),
  yourEngagmentDomains: faker.helpers.arrayElements(
    Object.values(EngagmentDomains),
  ),

  emailApiCalled: false,
  broadcastChannel: faker.helpers.arrayElement(Object.values(BroadcastChannel)),
  email: Date.now() + faker.internet.email(),
  ...data,
  surveyId,
});
