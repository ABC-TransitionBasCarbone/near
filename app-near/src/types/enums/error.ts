export enum ErrorCode {
  WRONG_SURVEY_PHASE = "WRONG_SURVEY_PHASE",
  MISSING_SURVEY_PHASE = "MISSING_SURVEY_PHASE",
  SU_NOT_COMPUTED = "SU_NOT_COMPUTED",
  SU_SEND_EMAIL = "SU_SEND_EMAIL",
  UNEXPECTED_ERROR = "UNEXPECTED_ERROR",
  UNEXPECTED_COMPUTE_SU_ERROR = "UNEXPECTED_COMPUTE_SU_ERROR",
  MISSING_SURVEY_NAME = "MISSING_SURVEY_NAME",
  WRONG_SURVEY_NAME = "WRONG_SURVEY_NAME",
  UNEXPECTED_NGCFORM = "UNEXPECTED_NGCFORM",
  UNEXPECTED_NGCFORM_EMAIL = "UNEXPECTED_NGCFORM_EMAIL",
  NEIGHBORHOOD_NOT_FOUND = "NEIGHBORHOOD_NOT_FOUND",
  MISSING_FORM_ID = "MISSING_FORM_ID",
  WRONG_FORM_ID = "WRONG_FORM_ID",
  WRONG_SIGNATURE = "WRONG_SIGNATURE",
  SU_NOT_FOUND = "SU_NOT_FOUND",
  WRONG_BARYCENTRE_DATA = "WRONG_BARYCENTRE_DATA",
}

export const errorCodeMapper: Partial<Record<ErrorCode, string>> = {
  [ErrorCode.MISSING_SURVEY_PHASE]:
    "Vous n'êtes pas dans la bonne étape pour valider cette phase.",
  [ErrorCode.SU_NOT_COMPUTED]:
    "Vous devez d'abord calculter les sphères d'usages.",
  [ErrorCode.SU_SEND_EMAIL]:
    "Les emails n'ont pas été correctement envoyés, veuillez vous rapprocher du support pour déterminer l'impact.",
  [ErrorCode.WRONG_SURVEY_PHASE]:
    "Votre sondage n'est pas à une phase lui permettant de réaliser cette action.",
  [ErrorCode.UNEXPECTED_COMPUTE_SU_ERROR]:
    "Une erreur inattendue pour computeSus s'est produite",
};
