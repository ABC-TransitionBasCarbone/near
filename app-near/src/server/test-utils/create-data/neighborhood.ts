import { SurveyPhase } from "@prisma/client";
import { db } from "~/server/db";

export const createNeighborhood = async (surveyName: string) => {
  const survey = await db.survey.create({
    data: {
      name: surveyName,
      phase: SurveyPhase.STEP_2_SU_SURVERY,
      sampleTarget: 3,
    },
  });

  await db.$queryRawUnsafe(`CREATE OR REPLACE VIEW quartiers AS
      SELECT 
        ${survey.id} AS survey_id,
        ARRAY['75014', '75015', '75016']::text[] AS iris_selectors,
        1100::double precision AS population_sum_with_under_fifteen,
        1000::double precision AS population_sum,
        180::double precision AS p21_pop1529_sum,
        230::double precision AS p21_pop3044_sum,
        170::double precision AS p21_pop4559_sum,
        200::double precision AS p21_pop6074_sum,
        220::double precision AS p21_pop75p_sum,
        220::double precision AS c21_pop15p_cs1_sum,
        110::double precision AS c21_pop15p_cs2_sum,
        180::double precision AS c21_pop15p_cs3_sum,
        150::double precision AS c21_pop15p_cs4_sum,
        50::double precision AS c21_pop15p_cs5_sum,
        120::double precision AS c21_pop15p_cs6_sum,
        80::double precision AS c21_pop15p_cs7_sum,
        90::double precision AS c21_pop15p_cs8_sum,
        560::double precision AS population_femme_sum,
        440::double precision AS population_homme_sum,
        0::double precision AS population_sum_threshold_3p,
        0::double precision AS population_sum_threshold_4p,
        0::double precision AS population_sum_threshold_5p,
        0::double precision AS population_sum_threshold_4_5p
        ;`);

  return survey;
};
