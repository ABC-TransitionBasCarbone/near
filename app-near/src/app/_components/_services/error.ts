import { TRPCClientError } from "@trpc/client";
import { type ErrorCode, errorCodeMapper } from "~/types/enums/error";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getErrorValue = (e: any): string => {
  let value = "Une erreur s'est produite veuillez r√©-essayer plus tard";
  if (e instanceof TRPCClientError) {
    value = errorCodeMapper[e.message as ErrorCode];
  }

  return value;
};
