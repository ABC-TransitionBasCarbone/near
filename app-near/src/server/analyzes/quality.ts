import { allSurveyTypes, SurveyType } from "~/types/enums/survey";
import { db } from "../db";
import { TRPCError } from "@trpc/server";
import { ErrorCode } from "~/types/enums/error";
import {
  type QualityStatisticsWithPopulation,
  type QualityStatistics,
} from "~/types/QualityStatistics";

const surveyTypeMapper: Record<
  SurveyType,
  | typeof db.carbonFootprintAnswer
  | typeof db.wayOfLifeAnswer
  | typeof db.suAnswer
> = {
  [SurveyType.CARBON_FOOTPRINT]: db.carbonFootprintAnswer,
  [SurveyType.SU]: db.suAnswer,
  [SurveyType.WAY_OF_LIFE]: db.wayOfLifeAnswer,
};

const getQualityStatistics = async (
  type: SurveyType,
  surveyId: number,
  populationSize: number,
): Promise<QualityStatistics> => {
  const sampleSize = (await surveyTypeMapper[type].count({
    where: { surveyId },
  })) as number;

  const p = sampleSize / populationSize;

  const correctedMarginError =
    1.96 *
    Math.sqrt((p * (1 - p)) / sampleSize) *
    Math.sqrt((populationSize - sampleSize) / (populationSize - 1));

  return {
    type,
    sampleSize,
    correctedMarginError,
    confidence: 0.95,
  };
};

export const getAllQualityStatistics = async (
  surveyId: number,
): Promise<QualityStatisticsWithPopulation> => {
  const neighborhood = await db.quartier.findUnique({ where: { surveyId } });

  if (!neighborhood) {
    throw new TRPCError({
      code: "NOT_FOUND",
      message: ErrorCode.NEIGHBORHOOD_NOT_FOUND,
    });
  }

  const qualityStat = await Promise.all(
    allSurveyTypes.map((surveyType) =>
      getQualityStatistics(surveyType, surveyId, neighborhood.population_sum),
    ),
  );

  return {
    qualityStatistics: qualityStat,
    populationSum: neighborhood.population_sum,
    populationSumWith15: neighborhood.population_sum_with_under_fifteen,
  };
};
