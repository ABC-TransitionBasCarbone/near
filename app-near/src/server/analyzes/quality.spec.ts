import { TRPCError } from "@trpc/server";
import { getAllQualityStatistics } from "./quality";
import { ErrorCode } from "~/types/enums/error";
import { db } from "../db";
import { createNeighborhood } from "../test-utils/create-data/neighborhood";
import { buildSuAnswer } from "../test-utils/create-data/suAnswer";
import { type Survey } from "@prisma/client";
import { buildWayOfLifeAnswer } from "../test-utils/create-data/wayOfLifeAnswer";
import { buildCarbonFootprintAnswer } from "../test-utils/create-data/carbonFootprintAnswer";

describe("quality", () => {
  describe("getAllQualityStatistics", () => {
    let survey: Survey;
    const surveyName = "test-getAllQualityStatistics";
    beforeEach(async () => {
      await db.suAnswer.deleteMany().catch(() => null);
      await db.carbonFootprintAnswer.deleteMany().catch(() => null);
      await db.wayOfLifeAnswer.deleteMany().catch(() => null);
      await db.survey.delete({ where: { name: surveyName } }).catch(() => null);

      survey = await createNeighborhood(surveyName);

      await db.suAnswer.createMany({
        data: Array.from(Array(10).keys()).map(() => buildSuAnswer(survey.id)),
      });
      await db.wayOfLifeAnswer.createMany({
        data: Array.from(Array(20).keys()).map(() =>
          buildWayOfLifeAnswer(survey.id),
        ),
      });
      await db.carbonFootprintAnswer.createMany({
        data: Array.from(Array(30).keys()).map(() =>
          buildCarbonFootprintAnswer(survey.id),
        ),
      });
    });

    it("should throw 404 when neighorhood is not found", async () => {
      expect.assertions(2);
      try {
        await getAllQualityStatistics(32313131);
      } catch (error) {
        if (error instanceof TRPCError) {
          expect(error.code).toBe("NOT_FOUND");
          expect(error.message).toBe(ErrorCode.NEIGHBORHOOD_NOT_FOUND);
        }
      }
    });

    it("should return quality statistices", async () => {
      const result = await getAllQualityStatistics(survey.id);

      expect(result).toMatchSnapshot();
    });
  });
});
