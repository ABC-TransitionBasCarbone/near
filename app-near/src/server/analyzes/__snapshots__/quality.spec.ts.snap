// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quality getAllQualityStatistics should return quality statistices 1`] = `
[
  {
    "confidence": 1.96,
    "correctedMarginError": 0.06139153916527164,
    "sampleSize": 10,
    "type": "su",
  },
  {
    "confidence": 1.96,
    "correctedMarginError": 0.060771422608046674,
    "sampleSize": 20,
    "type": "way_of_life",
  },
  {
    "confidence": 1.96,
    "correctedMarginError": 0.0601513060508217,
    "sampleSize": 30,
    "type": "carbone_footprint",
  },
]
`;
